### ISSUE TITLE
ThrowsConstraint behavior inconsistent for async delegates 

### ISSUE THREAD

**Issue by appel1** [nunit/nunit#813]
ThrowsConstraint behaves differently depending on whether the exception is thrown directly by the method under test or returned as part of the resulting Task.

``` csharp
[TestFixture]
public class TestFixture
{
  [Test]
  public void TestFails()
  {
    Assert.That(() => ThrowsDirectly(), Throws.TypeOf<InvalidOperationException>());
  }

  [Test]
  public void TestSucceeds()
  {
    Assert.That(() => ThrowsAsync(), Throws.TypeOf<InvalidOperationException>());
  }

  private Task ThrowsDirectly()
  {
    throw new InvalidOperationException();
  }

  private async Task ThrowsAsync()
  {
    await Task.Yield();
    throw new InvalidOperationException();
  }
}
```

**Comment by CharliePoole** [comment]
This is closely related to #464

**Comment by spirit11** [comment]
ThrowsConstraint in version=3.6.1 still behaves differently for some cases:

```cs
[TestFixture]
public class Tests : AssertionHelper
{
    [Test]
    public void AsyncSuccess()
    {
        Expect(async () => await Task.FromException(new InvalidOperationException()), 
                Throws.Exception.TypeOf<InvalidOperationException>());
    }

    [Test]
    public void AsyncFail()
    {
        Expect(async () => await Task.FromException(new InvalidOperationException()), 
                Throws.Exception);
    }

    [Test]
    public void SyncSuccess()
    {
        Expect(() => { throw new InvalidOperationException(); }, 
                Throws.Exception.TypeOf<InvalidOperationException>());

        Expect(() => { throw new InvalidOperationException(); },
                 Throws.Exception);
    }
}
```

**Comment by mikkelbu** [comment]
As far as I can deduct from #1899 this will be solved in version 3.9.

**Comment by jnm2** [comment]
@mikkelbu is correct. #1899 was a duplicate of this issue and we published the fix with release 3.9.0 in November.

I'll close this issue as a duplicate of https://github.com/nunit/nunit/issues/1899 even though, in fairness, it should have been the other way around. 😃
